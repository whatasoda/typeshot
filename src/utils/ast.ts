import type ts from 'typescript';
import type { CodeStack } from './stack-tracking';

export const getSourceFileByStack = (stack: CodeStack, getSourceFile: (filename: string) => ts.SourceFile | null) => {
  const { filename } = stack;
  const sourceFile = getSourceFile(filename);
  if (sourceFile) {
    return sourceFile;
  } else {
    throw new Error('Source File Not Found: check tsconfig or typeshot options');
  }
};

export const getNodeByStack = <T extends ts.Node = ts.Node>(
  stack: CodeStack,
  sourceFile: ts.SourceFile,
  filterPath?: (node: ts.Node) => node is T,
) => {
  const { filename, line, col } = stack;
  if (filename !== sourceFile.fileName) throw new Error('Invalid Source File: filename unmatched');

  const lines = sourceFile.getLineStarts();
  if (line >= lines.length) throw new Error('Unexpected Line Information: something wrong happened');

  const pos = lines[line - 1] + col - 1;
  return getNodeByPosition(pos, sourceFile, filterPath);
};

export const getNodeByPosition = <T extends ts.Node = ts.Node>(
  pos: number,
  sourceFile: ts.SourceFile,
  filterPath?: (node: ts.Node) => node is T,
) => {
  const { statements } = sourceFile;
  if (statements[0].parent !== sourceFile) throw new Error('Make sure to turn setParentNodes true');

  let layer: readonly ts.Node[] = statements;
  const nodePath: T[] = [];
  while (layer.length) {
    LAYER_LOOP: for (let i = 0, length = layer.length; i < length; i++) {
      const node = layer[i];
      if (pos === node.getStart()) {
        if (!filterPath || filterPath(node)) {
          nodePath.push(node as T);
        }
        return { node, nodePath, sourceFile };
      }

      const nextNode: ts.Node | undefined = layer[i + 1];
      if (!nextNode || pos < nextNode.getStart()) {
        if (!filterPath || filterPath(node)) {
          nodePath.push(node as T);
        }
        layer = node.getChildren();
        break LAYER_LOOP;
      }
    }
  }
  throw new Error('Node Not Found: something wrong happened');
};

export const forEachChildDeep = (root: ts.Node, callback: (node: ts.Node) => true | void) => {
  const stack: ts.Node[] = [root];
  while (stack.length) {
    const node = stack.shift()!;
    if (callback(node) === true) {
      continue; // skip children
    } else {
      stack.unshift(...node.getChildren());
    }
  }
};
