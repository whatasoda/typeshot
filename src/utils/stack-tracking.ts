export interface CodeStack {
  composed: string;
  filename: string;
  line: number;
  col: number;
}

export const parseStack = (stack: string, depth: number): CodeStack => {
  let cursor = 0;
  for (let i = depth + 1; stack && i > 0; i--) {
    cursor = stack.indexOf('\n', cursor + 1);
  }
  cursor = (cursor === -1 ? stack.lastIndexOf('\n') : cursor) + 8 /* '____at_' */;
  const composed = stack
    .slice(cursor, (stack.indexOf('\n', cursor) + 1 || Infinity) - 1)
    .replace(STACK_PARENTHESES, '');

  if (!stack) throw new Error('Invalid Stack Information: invalid depth');
  const [filename, line, col] = composed.split(':');
  return { composed, filename, line: ~~line, col: ~~col };
};
const STACK_PARENTHESES = /(^.+\(|\)$)/g;

const target = Object.create(null) as { stack: string };
const cacheStore = new Map<string, CodeStack>();
export const withStackTracking = <T, U extends any[]>(func: (stack: CodeStack, ...args: U) => T) => {
  return (...args: U) => {
    Error.stackTraceLimit = 2;
    Error.captureStackTrace(target);
    Error.stackTraceLimit = 10;
    return func(cacheStore.get(target.stack) || parseStack(target.stack, 1), ...args);
  };
};
