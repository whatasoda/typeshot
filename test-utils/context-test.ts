import { getContext } from '../src/program/context';

(getContext() as any).test();

throw new Error('test error');
