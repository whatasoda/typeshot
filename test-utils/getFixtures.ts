import ts from 'typescript';
import path from 'path';

export const FIXTURE_DIR = path.resolve(__dirname, '../__fixtures__');

export const getFixtures = jest.fn((name: string, setParentNodes: boolean) => {
  const fileName = `${FIXTURE_DIR}/${name}.ts`;
  const sourceText = ts.sys.readFile(fileName);
  if (!sourceText) throw new Error(`fixture file '${name}' is not found`);
  return ts.createSourceFile(fileName, sourceText, ts.ScriptTarget.Latest, setParentNodes, ts.ScriptKind.TS);
});
