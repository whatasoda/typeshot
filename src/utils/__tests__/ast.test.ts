import ts from 'typescript';
import { getFixtures } from '../../../test-utils/getFixtures';
import { forEachChildDeep, getNodeByStack, getSourceFileByStack } from '../ast';
import { CodeStack } from '../stack-tracking';

const fixtrue = getFixtures('sample', true);

const stack = (line: number, col: number): CodeStack => ({
  composed: `${fixtrue.fileName}:${line}:${col}`,
  filename: fixtrue.fileName,
  line,
  col,
});

describe('getSourceFileByStack', () => {
  const testStack: CodeStack = { filename: 'testFileName', col: 0, line: 0, composed: 'testFileName:0:0' };
  it('works correctly', () => {
    const getSourceFile = jest.fn(() => fixtrue);
    const result = getSourceFileByStack(testStack, getSourceFile);

    expect(result).toBe(fixtrue);
    expect(getSourceFile).toBeCalledWith('testFileName');
  });

  it('throws error if source file is not found', () => {
    expect(() => getSourceFileByStack(testStack, () => null)).toThrowError();
  });
});

describe('getNodeByStack', () => {
  it('returns node and nodePath according to stack information', () => {
    const { node, nodePath, sourceFile } = getNodeByStack(stack(20, 28), fixtrue);
    expect(node.getText()).toBe('registerTypeDefinition');
    expect(nodePath[nodePath.length - 1]).toBe(node);
    expect(sourceFile).toBe(fixtrue);
  });

  it('filters nodePath if filterPath is specified', () => {
    const { node, nodePath, sourceFile } = getNodeByStack(stack(20, 28), fixtrue, ts.isCallExpression);
    expect(node.getText()).toBe('registerTypeDefinition');
    expect(sourceFile).toBe(fixtrue);

    const nearest = nodePath[nodePath.length - 1];
    expect(nearest).not.toBe(node);
    expect('\n' + nearest.getText() + '\n').toBe(`
typeshot.registerTypeDefinition((fields: FieldDefinition[], makeType) => {
  const acc = {};
  const resolvedIdSet = new Set<string>();

  fields.forEach(({ id, type, required }) => {
    if (resolvedIdSet.has(id)) throw new Error('Duplicated field id found!');

    const t = required
      ? makeType<{ [K in typeof id]-?: TypeMap[typeof type] } & string[]>({ id, type })
      : makeType<{ [K in typeof id]+?: TypeMap[typeof type] } & string[]>({ id, type });
    Object.assign(acc, t);
  });
  const aaa = 'aaa';
  (acc as any).aaa = makeType<number | typeof aaa>({ aaa });

  return acc;
})
`);
  });

  it('throws error if file name is not matched', () => {
    expect(() => {
      getNodeByStack({ col: 0, line: 0, filename: 'test', composed: 'test' }, fixtrue);
    }).toThrowError();
  });

  it('throws error if source file is not created with setParentNodes true', () => {
    const altFixture = getFixtures('sample', false);
    expect(() => {
      getNodeByStack(stack(20, 28), altFixture);
    }).toThrowError();
  });

  it('throws error if stack is not matched with source file', () => {
    expect(() => getNodeByStack(stack(1000, 10), fixtrue)).toThrowError();
    expect(() => getNodeByStack(stack(20, 29), fixtrue)).toThrowError();
  });
});

describe('forEachChildDeep', () => {
  it('works correctly', () => {
    const acc: string[] = [];
    forEachChildDeep(fixtrue.statements[2], (node) => {
      acc.push(node.getFullText());
    });
    expect(acc.join('\n' + '-'.repeat(20) + '\n')).toMatchSnapshot();
  });
  it('works correctly with returning true', () => {
    const acc: string[] = [];
    const root = fixtrue.getChildren()[0];
    forEachChildDeep(root, (node) => {
      if (node === root) return;
      acc.push(node.getFullText());
      return true;
    });
    expect(acc.join('\n' + '-'.repeat(20) + '\n')).toMatchSnapshot();
  });
});
