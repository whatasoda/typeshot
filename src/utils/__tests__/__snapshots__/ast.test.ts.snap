// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`forEachChildDeep works correctly 1`] = `
"

interface TypeMap {
  text: string;
  number: number;
}
--------------------


interface
--------------------
 TypeMap
--------------------
 {
--------------------

  text: string;
  number: number;
--------------------

  text: string;
--------------------

  text
--------------------
:
--------------------
 string
--------------------
;
--------------------

  number: number;
--------------------

  number
--------------------
:
--------------------
 number
--------------------
;
--------------------

}"
`;

exports[`forEachChildDeep works correctly with returning true 1`] = `
"import typeshot from 'typeshot';
--------------------


typeshot.openTrace\`
// GENERATED FILE -- DO NOT EDIT
\`;
--------------------


interface TypeMap {
  text: string;
  number: number;
}
--------------------


export interface FieldDefinition {
  id: string;
  type: keyof TypeMap;
  required: boolean;
}
--------------------


typeshot.closeTrace();
--------------------


const PropsType = typeshot.registerTypeDefinition((makeType, fields: FieldDefinition[]) => {
  const acc = {};
  const resolvedIdSet = new Set<string>();

  fields.forEach(({ id, type, required }) => {
    if (resolvedIdSet.has(id)) throw new Error('Duplicated field id found!');

    const t = required
      ? makeType<{ [K in typeof id]-?: TypeMap[typeof type] } & string[]>({ id, type })
      : makeType<{ [K in typeof id]+?: TypeMap[typeof type] } & string[]>({ id, type });
    Object.assign(acc, t);
  });
  const aaa = 'aaa';
  (acc as any).aaa = makeType<number | typeof aaa>({ aaa });

  return acc;
});
--------------------


const Props = PropsType([
  { id: 'foo', type: 'number', required: false },
  { id: 'bar', type: 'number', required: true },
  { id: 'baz', type: 'text', required: false },
  { id: 'qux', type: 'text', required: true },
]);
--------------------


typeshot.print\`
  export type Hoge = \${Props.literal()}
\`;
--------------------


typeshot.openTrace();
--------------------


// hogehoge
typeshot.closeTrace();"
`;
